components/director/internal/domain/formation/service.go

```diff
@@ -625,6 +636,29 @@ func (s *service) assign(ctx context.Context, tnt, objectID string, objectType g
 	return nil
 }
 
+func (s *service) unassign(ctx context.Context, tnt, objectID string, objectType graphql.FormationObjectType, formation *model.Formation) error {
+	switch objectType {
+	case graphql.FormationObjectTypeApplication:
+		if err := s.modifyAssignedFormations(ctx, tnt, objectID, formation.Name, objectTypeToLabelableObject(objectType), deleteFormation); err != nil {
+			return err
+		}
+	case graphql.FormationObjectTypeRuntime, graphql.FormationObjectTypeRuntimeContext:
+		if isFormationComingFromASA, err := s.asaEngine.IsFormationComingFromASA(ctx, objectID, formation.Name, objectType); err != nil {
+			return err
+		} else if isFormationComingFromASA {
+			return apperrors.NewCannotUnassignObjectComingFromASAError(objectID)
+		}
+
+		if err := s.modifyAssignedFormations(ctx, tnt, objectID, formation.Name, objectTypeToLabelableObject(objectType), deleteFormation); err != nil {
+			return err
+		}
+
+	default:
+		return nil
+	}
```

Comment:
It's to avoid problems with transactions. In the resynchronizing formations case we need to unassign the object if there are no formation assignments, but if we use `Unassign`, first we will have to mock the whole method and second, we will have the problem with `Unassign` opening its own transaction and seeing the now-deleted FA, resending the notifications (if there are any) and not unassigning it from the formation.
The error will be handled in the second switch - this one just handles with labeldef manipulation if it's applicable.